### DISCUSSION TITLE
Can't use .readinto on zlib.DecompIO: TypeError: object with buffer protocol required

### DISCUSSION THREAD

**Discussion by knuxify** [micropython/micropython#9291] [Using MicroPython] [answered]
The following code fails (only relevant bits left in):

```python
def open_dir(path):
    files = []

    for _file in os.listdir(path):
        file = path + '/' + _file
        if file.endswith('.bin'):
            os.remove(file)

    for _file in os.listdir(path):
        print(_file)
        filename = path + '/' + _file
        if filename.endswith('.zl'):
            file_out = filename.replace('.zl', '.bin')
            print(file_out)
            decompress_file(filename, file_out)
            files.append(file_out)

    (.....)

def decompress_file(in_path, out_path):
    with open(in_path, 'rb') as infile, open(out_path, 'wb') as outfile:
        decompio = zlib.DecompIO(infile)
        decompio.readinto(outfile)
```

Returning the following traceback:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "main.py", line 85, in <module>
  File "main.py", line 50, in open_dir
  File "main.py", line 65, in decompress_file
TypeError: object with buffer protocol required
```

The code here fails on the `.readinto` line.

It seems that the usual fix for this problem is to "use strings/bytes instead of passed object", but this doesn't make sense here as all of these objects should have the buffer protocol implemented. I checked the types for `infile`, `outfile` and `decompio` and they all seem to match what's expected (`FileIO`, `FileIO`, `DecompIO`).

It's worth noting that I can use the `.read` method just fine (although it is quite slow, which I'd rather avoid...), and .readinto on DecompIO works when reading directly from an input file into a bytearray.

Am I doing something wrong, or did I hit some bizzare bug? Could it be an out-of-memory error in disguise?

I'm using a Raspberry Pi Pico RP2040, with `rp2-pico-20220909-unstable-v1.19.1-389-g4903e48e3.uf2`.

**Comment by dpgeorge** [comment, marked as answer]
> It seems that the usual fix for this problem is to "use strings/bytes instead of passed object"

Yes that is the correct way to do it, create a temporary `bytearray` to buffer the data from one stream to another.

> but this doesn't make sense here as all of these objects should have the buffer protocol implemented

There is a difference between "buffer protocol" and "stream protocol".  The former is an in-memory object like `str`, `bytes`, `bytearray` (mutable), `memoryview` and some other things.  The latter is an object that has `read`/`write` methods (essentially inherits from `IOBase`).

I don't think CPython even allows reading directly from one file/stream into another?
